File: clojure/beer-song/src/beer_song.clj
(ns beer-song
  (:require [clojure.pprint :refer [cl-format]]
            [clojure.string :as string]))

(def verse-format
  "~[No more~:;~:*~d~] bottle~:p of beer on the wall, ~:*~[no more~:;~:*~d~] bottle~:p of beer.
~:*~[Go to the store and buy some more, 99 bottles of beer on the wall~:;~
     Take ~[it~:;one~] down and pass it around, ~:*~[no more~:;~:*~d~] bottle~:p of beer on the wall~].~%")


(defn verse
  "Returns the nth verse of the song."
  [num]
  (cl-format nil verse-format num (dec num)))

(defn sing
  "Given a start and an optional end, returns all verses in this interval. If
  end is not given, the whole song from start is sung."
  ([start] (sing start 0))
  ([start end]
   (->>
    (range start (dec end) -1)
    (map verse)
    (string/join "\n"))))

File: ruby/beer-song/beer_song.rb
class BeerSong
  class << self
    def recite(start, size)
      start.downto(start - size + 1).map { verse _1 }.join("\n")
    end

    def verse(bottles)
      case bottles
      when 0
        <<~VERSE
          No more bottles of beer on the wall, no more bottles of beer.
          Go to the store and buy some more, 99 bottles of beer on the wall.
        VERSE
      when 1
        <<~VERSE
          1 bottle of beer on the wall, 1 bottle of beer.
          Take it down and pass it around, no more bottles of beer on the wall.
        VERSE
      when 2
        <<~VERSE
          2 bottles of beer on the wall, 2 bottles of beer.
          Take one down and pass it around, 1 bottle of beer on the wall.
        VERSE
      else
        <<~VERSE
          #{bottles} bottles of beer on the wall, #{bottles} bottles of beer.
          Take one down and pass it around, #{bottles - 1} bottles of beer on the wall.
        VERSE
      end
    end
  end
end
